### PR TITLE
chore(ci): itr on ddtrace

### PR DESCRIPTION
**Pull request by gnufede**
Enables Intelligent Test Runner on ddtrace, allowing to skip non-required tests in the CI.

Depends on:
- [x] #6543 
- [x] #6545 
- [x] #6546  
- [x] #6547 
- [x] #6548 
- [x] #6565 

## Checklist

- [x] Change(s) are motivated and described in the PR description.
- [x] Testing strategy is described if automated tests are not included in the PR.
- [x] Risk is outlined (performance impact, potential for breakage, maintainability, etc).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html) are followed. If no release note is required, add label `changelog/no-changelog`.
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/)).
- [x] Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))

## Reviewer Checklist

- [x] Title is accurate.
- [x] No unnecessary changes are introduced.
- [x] Description motivates each change.
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Testing strategy adequately addresses listed risk(s).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] Release note makes sense to a user of the library.
- [x] Reviewer has explicitly acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment.
- [x] Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)

### CODE REVIEW

File: .circleci/config.templ.yml [hunk 1]
```diff
@@ -36,6 +36,7 @@ machine_executor: &machine_executor
     # https://support.circleci.com/hc/en-us/articles/360045268074-Build-Fails-with-Too-long-with-no-output-exceeded-10m0s-context-deadline-exceeded-
     - PYTHONUNBUFFERED: 1
   steps:
+    - run: sudo apt-get update && sudo DEBIAN_FRONTEND=noninteractive apt-get install -o Dpkg::Options::="--force-confdef" -o Dpkg::Options::="--force-confold" --force-yes -y --no-install-recommends openssh-client git
```
Comment (brettlangdon) [inline .circleci/config.templ.yml:39]: should we be using a different base image? we are using ubuntu-2004, wonder if we should be using a more recent version?

this is likely an aside and not a requirement to change.

File: .coveragerc [hunk 2]
```diff
@@ -1,5 +1,6 @@
 [run]
 omit = ddtrace/vendor/*
+parallel = true
```
Comment (brettlangdon) [inline .coveragerc:3]: what does this do?

File: riotfile.py [hunk 3]
```diff
@@ -129,7 +132,7 @@ def select_pys(min_version=MIN_PYTHON_VERSION, max_version=MAX_PYTHON_VERSION):
         Venv(
             name="appsec",
             pys=select_pys(),
-            command="pytest {cmdargs} tests/appsec",
+            command="pytest --ddtrace {cmdargs} tests/appsec",
```
Comment (brettlangdon) [inline riotfile.py:135]: should we just add `--ddtrace` to the pytest config in `setup.cfg` ?

File: riotfile.py [hunk 4]
```diff
@@ -202,7 +205,7 @@ def select_pys(min_version=MIN_PYTHON_VERSION, max_version=MAX_PYTHON_VERSION):
         Venv(
             name="integration",
             # Enabling coverage for integration tests breaks certain tests in CI
-            command="pytest --no-cov {cmdargs} tests/integration/",
+            command="pytest --no-cov --ddtrace --ignore-glob='*civisibility*' {cmdargs} tests/integration/ && pytest --no-cov {cmdargs} tests/integration/test_integration_civisibility.py",  # noqa: E501
```
Comment (brettlangdon) [inline riotfile.py:208]: comment here would be nice, why do we need to break them up? I assume we cannot run with `--ddtrace` for civisibility integration tests?

File: riotfile.py (hunk 3 above)
Comment (brettlangdon) [reply riotfile.py:135]: I see below we are not using it for one scenario, are we able to selectively disable for a single run vs needing to enable everywhere?

Comment (P403n1x87) [inline riotfile.py:135]: Are there benefits/downsides for running this locally as well? If not, perhaps we can pass this option only via the CircleCI config

Comment (gnufede) [reply riotfile.py:135]: Yes, there are a few cases where it doesn't work (mainly pytest tests and similar): pytest, ci_visibility, pytest-bdd and asynctest

I can explore the option to create a `--no-ddtrace` option and see if it works.

Is it OK if I do that in a follow-up PR?

File: riotfile.py (hunk 4 above)
Comment (gnufede) [reply riotfile.py:208]: Yes, that's the reason, I can add a comment

File: riotfile.py (hunk 3 above)
Comment (gnufede) [reply riotfile.py:135]: We can run it locally as well, but it may not work properly for changes that are not committed, so maybe it's worth it to try to run it only in CircleCI

File: .circleci/config.templ.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (gnufede) [reply .circleci/config.templ.yml:39]: Yeah, it definitely should be better, since this is time consuming, but I wanted to keep the changes to the minimum here

File: .coveragerc (hunk 2 above)
Comment (gnufede) [reply .coveragerc:3]: This resolves an issue with coverage, since we're tracking the coverage per test with ITR, and there is another coverage instance running, there are conflicts while trying to persist the coverage file if we don't use this option.

File: riotfile.py (hunk 3 above)
Comment (gnufede) [reply riotfile.py:135]: How about passing a `--no-ddtrace` option locally to disable it?

Comment (P403n1x87) [reply riotfile.py]: > but it may not work properly for changes that are not committed

So the best benefit of ITR is in CI then? If so, personally I'd prefer to have the CLI option passed only in via the circleci config to minimise the impact on contributors wanting to run tests locally.

Comment (romainkomorndatadog) [reply riotfile.py]: There should be some benefit locally, too, though it requires `DD_API_KEY` and `DD_APPLICATION_KEY` env vars, and either setting up the service in on https://app.datadoghq.com/ci/settings/test-service to be enabled for the `DD_SERVICE` and `DD_ENV` default .

I think passing the CLI option through the CircleCI is the safe option but, on the other hand, it might be nice to get a bit of "forced" adoption.

File: setup.cfg [hunk 5]
```diff
@@ -40,6 +40,7 @@ rst-directives = py:data
 [tool:pytest]
 # --cov-report is intentionally empty else pytest-cov will default to generating a report
 addopts =
+  --ddtrace
```
Comment (gnufede) [inline setup.cfg:43]: @brettlangdon @P403n1x87 Should we do some trick to have this line only in the CI? Something like a separate setup.cfg for CI than for local development?

Comment (gnufede) [reply setup.cfg]: OK, fixed by changing the command in the CI template file

File: ddtrace/contrib/pytest/plugin.py [hunk 6]
```diff
@@ -62,7 +62,7 @@ def encode_test_parameter(parameter):
 
 def is_enabled(config):
     """Check if the ddtrace plugin is enabled."""
-    return config.getoption("ddtrace") or config.getini("ddtrace")
+    return (config.getoption("ddtrace") or config.getini("ddtrace")) and not config.getoption("no-ddtrace")
```
Comment (brettlangdon) [inline ddtrace/contrib/pytest/plugin.py:65]: sorry to do it to you, but would you be willing to make this it's own PR.

we probably want to add a release note as a feature here for `--no-ddtrace` since we are modifying the pytest plugin that customers also use.

File: riotfile.py [hunk 7]
```diff
@@ -1368,12 +1373,12 @@ def select_pys(min_version=MIN_PYTHON_VERSION, max_version=MAX_PYTHON_VERSION):
...
         Venv(
             name="boto",
-            command="pytest {cmdargs} tests/contrib/boto",
+            command="pytest --no-ddtrace {cmdargs} tests/contrib/boto",
             venvs=[Venv(pys=select_pys(max_version="3.6"), pkgs={"boto": latest, "moto": "<1.0.0"})],
         ),
         Venv(
             name="botocore",
-            command="pytest {cmdargs} tests/contrib/botocore",
+            command="pytest --no-ddtrace {cmdargs} tests/contrib/botocore",
```
Comment (brettlangdon) [inline riotfile.py:1381]: can we document these exceptions (code comments are fine), like, will this never work with `--ddtrace`, or is this something we are just saving as a todo to fix later?

Comment (gnufede) [reply riotfile.py:1381]: Actually I think this one should work, but I got some errors that I wasn't sure if they were failing in other branches, like, but maybe it's actually flaky?:
https://app.circleci.com/pipelines/github/DataDog/dd-trace-py/42049/workflows/7cf2183b-9f81-47eb-ab70-78a7886fa2dd/jobs/2805599/parallel-runs/3?filterBy=FAILED&invite=true#step-108-166
https://app.circleci.com/pipelines/github/DataDog/dd-trace-py/41887/workflows/c1231f57-cb36-4639-b42e-9d599b83d9cc/jobs/2797712?invite=true#step-108-141

Comment (Yun-Kim) [reply riotfile.py:1381]: I believe those were flaky tests that should be fixed by now? Can you try re-running without ITR disabled to verify?

**Review by mabdinur — COMMENTED** [review]
A few tests suites are failing: https://app.circleci.com/pipelines/github/DataDog/dd-trace-py/42484/workflows/2e21ac1d-c61e-44bb-bfed-2a91fff41bfc

**Review by mabdinur — APPROVED** [review]
